### PR TITLE
Interface CAMP with PartMC

### PR DESCRIPTION
**Pull request by jcurtis2**
Fixed PartMC to be able to use CAMP externally rather than within the PartMC repo itself as we decided to separate PartMC and CAMP development ([https://github.com/open-atmos/camp](https://github.com/open-atmos/camp)) from one another.

PartMC uses CAMP provided library and modules similar to the procedure used for MOSAIC.

* [x] Add plotting scripts to 6_camp scenario

### CODE REVIEW

File: scenarios/6_camp/monarch_mod37/cb05_mechanism.json.orig [hunk 1]
```diff
@@ -0,0 +1,3155 @@
+{ "pmc-data" : [
```
Comment (mattldawson) [inline scenarios/6_camp/monarch_mod37/cb05_mechanism.json.orig:1]: is this file needed?

File: scenarios/6_camp/monarch_mod37/cb05_mechanism.noR142R143.json [hunk 2]
```diff
@@ -0,0 +1,3112 @@
+{ "camp-data" : [
```
Comment (mattldawson) [inline scenarios/6_camp/monarch_mod37/cb05_mechanism.noR142R143.json:1]: Is this the mechanism we used in the paper? Or was it `cb05_mechanism.json`? I would suggest only including one and maybe naming it `cb05_mechanism.json`

**Review by mattldawson — DISMISSED** [review]
worked great for me! just suggested removing a couple configuration files that aren't used (I think)

File: src/aero_data.F90 [hunk 3]
```diff
@@ -332,6 +358,28 @@ subroutine aero_data_set_mosaic_map(aero_data)
...
+    !> Aerosol data.
+    class(aero_data_t), intent(in) :: aero_data
+    !> Computational particle index (1...aero_state_t%n_part)
+    integer, intent(in) :: i_part
+    !> Aerosol species index in aero_particle_t%vol(:) array
+    integer, intent(in) :: i_spec
+
+#ifdef PMC_USE_CAMP
+    call assert(106669451, allocated(aero_data%camp_particle_spec_id))
+    call assert(278731889, aero_data%camp_particle_state_size .ge. 0)
```
[the remark this answers is not in the document]
Comment (mwest1066) [reply src/aero_data.F90:374]: I changed all of these.

File: Dockerfile.camp [hunk 4]
```diff
@@ -0,0 +1,88 @@
+FROM fedora:27
```
[the remark this answers is not in the document]
Comment (mwest1066) [reply Dockerfile.camp]: Excellent!

[the remark this answers is not in the document]
Comment (mwest1066) [reply Dockerfile.camp]: Just to make sure I'm understanding, are we now using the fixed version of CAMP?

[the remark this answers is not in the document]
Comment (jcurtis2) [reply Dockerfile.camp]: Yep - switched to using v1.0.0-gamma

[the remark this answers is not in the document]
Comment (mwest1066) [reply Dockerfile.camp]: Excellent, thanks for confirming.

File: Dockerfile [hunk 5]
```diff
@@ -15,5 +15,11 @@ COPY . /partmc/
 
 RUN mkdir /build \
     && cd /build \
-    && cmake /partmc \
+    && cmake -D CMAKE_BUILD_TYPE=release \
+             -D CMAKE_C_FLAGS_DEBUG="-g" \
+             -D CMAKE_Fortran_FLAGS_DEBUG="-g" \
+             -D CMAKE_C_FLAGS_RELEASE="-O2 -g -Werror -Wall -Wextra" \
+             -D CMAKE_Fortran_FLAGS_RELEASE="-O2 -g -Werror -fimplicit-none -Wall -Wextra -Wconversion -Wunderflow -Wimplicit-interface -Wno-compare-reals -Wno-unused -Wno-unused-parameter -Wno-unused-dummy-argument -fbounds-check" \
+             -D ENABLE_GSL:BOOL=TRUE \
```
Comment (mwest1066) [inline Dockerfile:23]: @jcurtis2 can we also turn on sundials and other standard options here?

File: Dockerfile.camp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    && export CAMP_HOME="/camp-1.0.0-gamma/build/" \
+    && cmake -D CMAKE_BUILD_TYPE=release \
+             -D CMAKE_C_FLAGS_DEBUG="-g" \
+             -D CMAKE_Fortran_FLAGS_DEBUG="-g" \
+             -D CMAKE_C_FLAGS_RELEASE="-O2 -g -Werror -Wall -Wextra" \
+             -D CMAKE_Fortran_FLAGS_RELEASE="-O2 -g -Werror -fimplicit-none -Wall -Wextra -Wconversion -Wunderflow -Wimplicit-interface -Wno-compare-reals -Wno-unused -Wno-unused-parameter -Wno-unused-dummy-argument -fbounds-check" \
+             -D ENABLE_GSL:BOOL=TRUE \
+             -D ENABLE_CAMP=TRUE \
+             -D CAMP_LIB=/camp-1.0.0-gamma/build/lib/libcamp.so \
+             -D CAMP_INCLUDE_DIR=/camp-1.0.0-gamma/build/include/ \
```
Comment (mwest1066) [inline Dockerfile.camp:88]: Why are these two lines necessary? They should be auto-discovered from `CAMP_HOME`?

File: Dockerfile.camp [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+# NOTE: Modify .dockerignore to whitelist files/directories to copy.
+COPY . /partmc/
+
+RUN mkdir /build \
+    && cd /build \
+    && export CAMP_HOME="/camp-1.0.0-gamma/build/" \
+    && cmake -D CMAKE_BUILD_TYPE=release \
+             -D CMAKE_C_FLAGS_DEBUG="-g" \
+             -D CMAKE_Fortran_FLAGS_DEBUG="-g" \
```
Comment (mwest1066) [inline Dockerfile.camp:82]: We are setting `DEBUG` flags but building for `release`. This isn't necessarily bad, but is there a reason we are doing it?

File: src/aero_data.F90 [hunk 8]
```diff
@@ -332,6 +356,28 @@ subroutine aero_data_set_mosaic_map(aero_data)
 
   end subroutine aero_data_set_mosaic_map
 
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+  !> Get the index on the CAMP state array for a specified species and
+  !! computational particle
+  integer function camp_spec_id(aero_data, i_part, i_spec)
```
Comment (mwest1066) [inline src/aero_data.F90:363]: @jcurtis2 when I turned on the extra warnings, this function broke the non-CAMP test build. Can you figure out how to fix it?

File: src/run_part.F90 [hunk 9]
```diff
@@ -103,15 +113,19 @@ module pmc_run_part
...
        gas_state, run_part_opt)
-
+#endif
     !> Environment state.
     type(scenario_t), intent(in) :: scenario
     !> Environment state.
     type(env_state_t), intent(inout) :: env_state
     !> Aerosol data.
-    type(aero_data_t), intent(in) :: aero_data
+    type(aero_data_t), intent(inout) :: aero_data
```
[the remark this answers is not in the document]
Comment (mwest1066) [reply src/run_part.F90:128]: I haven't yet had time to carefully trace out exactly what this would look like if we move it to `aero_state`. I tend to think that's the correct solution. I don't mind introducing a pointer.

Do you want to take a go at moving this to `aero_state`? If you push a single commit that does only this change, then we can revert it if it turns out to be too horrible.

[the remark this answers is not in the document]
Comment (jcurtis2) [reply src/run_part.F90:128]: Yep, I'll move it to aero_state and see what happens!

File: src/aero_data.F90 (hunk 8 above)
Comment (jcurtis2) [reply src/aero_data.F90:363]: Fixed this by returning a value of 0 if this code is compiled without CAMP support.

File: Dockerfile (hunk 5 above)
Comment (jcurtis2) [reply Dockerfile:23]: We have to address #33 for our condensation code to work with newer versions of SUNDIALS.

**Review by mwest1066 — APPROVED** [review]

File: Dockerfile (hunk 5 above)
Comment (slayoo) [reply Dockerfile:21]: https://github.com/compdyn/partmc/pull/167
